src/nvoptix_callbacks.c

```diff
@@ -0,0 +1,105 @@
+#include "windef.h"
+#include "winbase.h"
+#include "wine/debug.h"
+
+WINE_DEFAULT_DEBUG_CHANNEL(nvoptix);
+
+#include <stdint.h>
+#include <stdlib.h>
+#include <pthread.h>
+
+#include "nvoptix.h"
+
+pthread_rwlock_t callbacks_lock;
+struct callback_t *callbacks = NULL;
+static ptrdiff_t callbacks_count = 0;
+
+_Bool callbacks_enabled(void)
+{
+    static int enabled = -1;
+
+    if (enabled == -1)
+    {
+        char *env = getenv("WINE_NVOPTIX_CALLBACKS");
+
+        if (env)
+        {
+            enabled = atoi(env) ? 1 : 0;
+
+            if (!enabled) WARN("Log callbacks disabled\n");
+        }
+        else
+        {
+            enabled = 1;
+        }
+    }
+
+    return enabled;
+}
+
+void *wrap_callback(OptixLogCallback func, void *data)
+{
+    if (pthread_rwlock_wrlock(&callbacks_lock))
+    {
+        ERR("(%p, %p): Failed to acquire writer lock\n", func, data);
+        return (void*)~(ptrdiff_t)0;
+    }
+
+    struct callback_t *new_callbacks = reallocarray(
+        callbacks, callbacks_count + 1, sizeof(struct callback_t));
+
+    if (!new_callbacks)
+    {
+        ERR("(%p, %p): Failed to reallocate callbacks array\n", func, data);
+
+        if (pthread_rwlock_unlock(&callbacks_lock))
+            ERR("(%p, %p): Failed to release writer lock\n", func, data);
+
+        return (void*)~(ptrdiff_t)0;
+    }
+
+    callbacks = new_callbacks;
+
+    struct callback_t *callback = &callbacks[callbacks_count];
+
+    *(void**)&callback->func = func;
+    callback->data = data;
+
+    ptrdiff_t offset = callbacks_count++;
+
+    if (pthread_rwlock_unlock(&callbacks_lock))
+        ERR("(%p, %p): Failed to release writer lock\n", func, data);
+
+    TRACE("(%p, %p) = %td\n", func, data, offset);
+
+    return (void*)offset;
+}
+
+void log_callback(unsigned int level, const char *tag, const char *message, void *cbdata)
+{
+    TRACE("(%u, %s, %s, %p)", level, tag, message, cbdata);
+
```

Comment:
Seems something bad happens if you try to TRACE the message... Using %p to just print the pointer seems to not crash.

`TRACE("(%u, %s, %p, %p)", level, tag, message, cbdata);`

Reply:
It could be some `string` limitation in Wine for debug (TRACE) messages perhaps, and it causes some sort of overload when extra info is printed in the message string when you get a cache miss? Not sure it is really worth pursuing much further, since either running with `WINEDEBUG=-all` or just printing the ptr to the message fixed the crash.

Reply:
I have a feeling this is related to messages being too long to fit in Wine's internal debug printf buffer. I believe wine-mono had similar issues in the past. Will fix.

Reply:
Also I just added the missing newline at the end :sweat_smile: